diff --git a/third-party/CMakeLists.txt b/third-party/CMakeLists.txt
index a185e2e..b417615 100644
--- a/third-party/CMakeLists.txt
+++ b/third-party/CMakeLists.txt
@@ -74,11 +74,14 @@ if(CPYTHON)
   set(LIBPYTHONNAME libpython3.8)
   set(LIBPYTHONNAMETAG v3.8.10)
   set(LIBPYTHONPIP pip3.8)
+  set(LDFLAGS "-Wl,-rpath=${CMAKE_INSTALL_PREFIX}/lib")
+
   EXTERNALPROJECT_ADD(
     ${LIBPYTHONNAME}
     PREFIX ${LIBPYTHONNAME}
     SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/cpython
-    CONFIGURE_COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/cpython/configure --prefix=${CMAKE_INSTALL_PREFIX} --enable-shared LDFLAGS=-Wl,-rpath=${CMAKE_INSTALL_PREFIX}/lib > /dev/null
+    INSTALL_DIR ${CMAKE_INSTALL_PREFIX}
+    CONFIGURE_COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/cpython/configure --prefix=<INSTALL_DIR> --enable-shared LDFLAGS=${LDFLAGS} > /dev/null
     BUILD_COMMAND ${CMAKE_MAKE_PROGRAM} -j${N} > /dev/null
     INSTALL_COMMAND ${CMAKE_MAKE_PROGRAM} altinstall > /dev/null
     COMMAND cd ${CMAKE_CURRENT_SOURCE_DIR}/pip_packages &&
